source/extensions/filters/common/ratelimit/ratelimit_impl.cc

```diff
@@ -105,15 +105,20 @@ void GrpcClientImpl::onSuccess(
 
   DescriptorStatusListPtr descriptor_statuses = std::make_unique<DescriptorStatusList>(
       response->statuses().begin(), response->statuses().end());
+  // TODO(esmet): This dynamic metadata copy is probably unnecessary, but let's just following the
+  // existing pattern of copying parameters over as unique pointers for now.
+  DynamicMetadataPtr dynamic_metadata =
+      std::make_unique<ProtobufWkt::Struct>(response->dynamic_metadata());
```

Comment:
We should avoid allocating here if there is no metadata which will be most cases. Can you guard this with `has_dynamic_metadata()`?

Reply:
Yeah that's fair. By doing that, we probably don't even need to optimize away the copy when there is metadata, since the common case is still zero copy/alloc.

So, I'll add the guard and remove the comment.